app/controllers/vouchers_controller.rb

```diff
@@ -0,0 +1,11 @@
+class VouchersController < ApplicationController
+  before_action :authenticate_retailer!
+  load_and_authorize_resource
+  def index
+    @vouchers
+  end
+
+  def show
+    @voucher
```

Comment:
This, as the one just above, automaticaly load the voucher based on the url. For ex `/voucher/17` will load the Voucher with the id 17 automatically. It works only if the voucher concerned belongs to the retailer logged in, otherwise, a "access denied" page is triggered